### ISSUE TITLE
Surprising behavior providing param overrides

### ISSUE THREAD

**Issue by johaywood** [oauth-xx/oauth2#352]
I'm running into an interesting scenario -- not sure it's quite a bug, but at least "surprising" -- where providing a set of params to `OAuth2::AccessToken#refresh!` that includes overrides to the `client_id` and `client_secret`. When merging these into the default params provided by `OAuth2::Authenticatior#apply_params_auth` the resulting request includes duplicate keys for `client_id` and `client_secret` in the request body. It looks like this same scenario would be possible in other cases where overrides can be supplied, this is just the only one I ran into and have tested.

This appears to be due to the fact I'm providing my overrides with symbol keys and the defaults are string keys. In Ruby allowing both keys in a hash is totally expected since keys can be any object; however, in the context of this gem it's likely not the intention to have multiple keys with the same semantic meaning.

This scenario isn't likely to be caught testing with something like Webmock to assert that there are no duplicate keys in a request body. When deserializing the query string the keys will be de-duped since we've lost the idea of string vs symbol at that point and the test will pass. Plus for most people this would probably fall under the "don't test the library, test your code" umbrella.

I'm curious about anyone else's thoughts on if this should be considered a bug and, if so, have a discussion about the best approach to fixing it.

**Comment by pboling** [comment]
Having URL parameters that are identical and pass values under the same key is a feature that many systems rely on for passing of a list of data, just on the internet generally.  This does not appear to be an intentional or useful example of that paradigm.  I would say this might be a bug.

A ruby hash converted to params can't directly result in reuse of the same parameter more than 2x, and there only due to the string / symbol dichotomy.  I wonder if there are OAuth use cases that we should support with multiple identical parameters.

**Comment by johaywood** [comment]
I'm sure there definitely are some legitimate use cases out there for list params, and those are supported by Faraday in a couple different ways depending on your use case ([example 1](https://github.com/lostisland/faraday/blob/565b2a49af3ddd01e5258fcd864247a2a8cc786e/test/request_middleware_test.rb#L32) [example 2](https://github.com/lostisland/faraday/blob/565b2a49af3ddd01e5258fcd864247a2a8cc786e/test/request_middleware_test.rb#L51)), though I haven't actually run into a scenario I've needed to use either of these in conjunction with the OAuth2 gem.

If I were supplying a params hash that's being merged into a default hash, my expectation would be that my hash would override the default values, and if I needed to supply a list query param I definitely wouldn't want to figure out whether I should supply a string or symbol and then only be limited to 2 values.

That said, now that I've convinced myself this is a bug, it's definitely an edge case that the average user of this gem likely won't experience, so I'd consider it a "nice-to-have". If I get some time I'll try to put a PR together to see if there's a good way to standardize the behavior globally.
